crates/sui-light-client/src/main.rs

```diff
@@ -0,0 +1,680 @@
+
+    // Check the signatures of all checkpoints
+    // And download any missing ones
+
+    let mut prev_committee = genesis_committee;
+    for ckp_id in &checkpoints_list.checkpoints {
+        // check if there is a file with this name ckp_id.yaml in the checkpoint_summary_dir
+        let mut checkpoint_path = config.checkpoint_summary_dir.clone();
+        checkpoint_path.push(format!("{}.yaml", ckp_id));
+
+        // If file exists read the file otherwise download it from the server
+        let summary = if checkpoint_path.exists() {
+            read_checkpoint(config, *ckp_id)?
+        } else {
+            // Download the checkpoint from the server
+            let summary = download_checkpoint_summary(config, *ckp_id).await?;
+            summary.clone().verify(&prev_committee)?;
+            // Write the checkpoint summary to a file
+            write_checkpoint(config, &summary)?;
+            summary
+        };
+
+        // Print the id of the checkpoint and the epoch number
+        println!(
+            "Epoch: {} Checkpoint ID: {}",
+            summary.epoch(),
+            summary.digest()
+        );
+
+        // Extract the new committee information
+        if let Some(EndOfEpochData {
+            next_epoch_committee,
+            ..
+        }) = &summary.end_of_epoch_data
+        {
+            let next_committee = next_epoch_committee.iter().cloned().collect();
+            prev_committee = Committee::new(summary.epoch().saturating_add(1), next_committee);
+        } else {
+            return Err(anyhow!(
+                "Expected all checkpoints to be end-of-epoch checkpoints"
+            ));
+        }
+    }
+
+    Ok(())
+}
+
+async fn get_full_checkpoint(config: &Config, seq: u64) -> anyhow::Result<CheckpointData> {
+    // Downloading the checkpoint from the server
+    let client: Client = Client::new(config.rest_url());
+    let full_checkpoint = client.get_full_checkpoint(seq).await?;
+
+    Ok(full_checkpoint)
+}
+
+fn extract_verified_effects_and_events(
+    checkpoint: &CheckpointData,
+    committee: &Committee,
+    tid: TransactionDigest,
+) -> anyhow::Result<(TransactionEffects, Option<TransactionEvents>)> {
+    let summary = &checkpoint.checkpoint_summary;
+
+    // Verify the checkpoint summary using the committee
+    summary.verify_with_contents(committee, Some(&checkpoint.checkpoint_contents))?;
+
+    // Check the validity of the transaction
+    let contents = &checkpoint.checkpoint_contents;
+    let (matching_tx, _) = checkpoint
+        .transactions
+        .iter()
+        .zip(contents.iter())
+        // Note that we get the digest of the effects to ensure this is
+        // indeed the correct effects that are authenticated in the contents.
+        .find(|(tx, digest)| {
+            tx.effects.execution_digests() == **digest && digest.transaction == tid
+        })
+        .ok_or(anyhow!("Transaction not found in checkpoint contents"))?;
```

Comment:
iiuc we do not verify user signature for the matching_tx, correct?

Reply:
We do not do it here, and I think we also do not do it upon full node re-execution.